Refactor the construction of the initial environment (fixes #7841)

diml opened the pull request:
This PR is a work in progress, I'm opening it for early feedback and to make sure I'm approaching the problem the right way, as it is taking longer than I expected.

## Overview

This PR refactors the way include directories and handled. In particular, it allows modules implemented by external cmi files to shadow modules in the current environment. The main motivation is so that modules coming from include directories specified by the user have precedence over the implicit `open Stdlib`  performed by the compiler.

This is already the case, however the way it is implemented is not very elegant and doesn't work when incrementally adding include directories, such as in the toplevel.

## More details

In the current world, the environment is split in two:

1. a local in-memory environment
2. a mapping from module names to external cmi files, specified via a list of include directories

1 always have precedence over 2 which is only used as a fallback for modules that are not found in 1. In particular if we consider the following toplevel session:

```
# module X = struct end;;
# #directory "foo";;
```

It is impossible to refer to the compilation unit `X` stored in `foo/x.cmi` because the local `X` has precedence over external modules.

This PRs removes the fallback  to 2 and replaces it by a new primitive that allows to add a persistent module to the environment. This means that adding an include directory becomes an operation that eagerly add names to the environment, shadowing existing ones. However, the cmi files are still loaded lazily.

If we reuse the DSL from #1513, this means the initial environment is now effectively constructed as follow:

```
env <- empty

env <- add_dir(env, stdlib_dir)
env <- open(env, Stdlib)

env <- add_dir(env, d1)
...
env <- add_dir(env, dn)

env <- open(env, M1)
...
env <- open(env, Mm)
```

Where `dX` are the include directories specified by the user via the `-I` option and `MX` are the implicitly opened modules specified via `-open`. In the toplevel, the `#directory <dir>` directive is immediate interpreted as: `env <- add_dir(env, dir)`. So if we consider the toplevel session above, after `#directory "foo"`, `X`  refers to `foo/x.cmi` rather than the local module `X`. This seems like a better behavior to me.

alainfrisch commented on toplevel/topdirs.ml:
```diff
@@ -83,7 +90,7 @@ let _ = add_directive "directory" (Directive_string dir_directory)
 let dir_remove_directory s =
   let d = expand_directory Config.standard_library s in
-  Config.load_path := List.filter (fun d' -> d' <> d) !Config.load_path;
+  if true then assert false;
```
What's the plan w.r.t. remove_directory?

diml replied:
I believe the following will give the same behavior as before: we look at all the visible modules in the environment and remove all the ones that are persistent and not yet loaded.

alainfrisch replied:
What do you mean?  The new code will always `assert false`, right?

diml replied:
Yh, I just haven't implemented it yet

alainfrisch replied:
Ah ok, I misunderstood your previous note.

alainfrisch commented on typing/env.ml:
```diff
@@ -2144,6 +2111,41 @@ let read_signature modname filename =
 
+let add_persistent_structure id env =
+  if not (Ident.persistent id) then invalid_arg "Env.add_persistent_structure";
+  if Ident.name id <> !current_unit then begin
```
(Nitpick: useless begin...end)

alainfrisch commented on typing/env.ml:
```diff
@@ -2144,6 +2111,41 @@ let read_signature modname filename =
+  Load_path.Dir.files dir
+  |> List.to_seq
+  |> Seq.filter_map unit_name_of_filename
+  |> String.Set.of_seq
```
(One should really have a List.filter_map....)

diml replied:
Indeed

dra27 replied:
We do have it - `Misc.Stdlib.List.filter_map`!

diml replied:
Indeed, I updated the code

alainfrisch commented on typing/env.ml:
```diff
@@ -2259,32 +2261,27 @@ let find_all_simple_list proj1 proj2 f lid env acc =
         acc
     | Some l ->
-      let p, desc = lookup_module_descr ~mark:true l env in
+        let p, desc = lookup_module_descr ~mark:true l env in
```
Please reindent the code below as well.

alainfrisch commented on typing/typemod.ml:
```diff
@@ -143,37 +143,56 @@ let type_open_ ?used_slot ?toplevel ovf env loc lid =
+    | Some m ->
+        (* Locate the directory that contains [m], adds the units it
+           contains to the environment and open [m] in the resulting
+           environment. *)
```
Can you explain the logic here?  Why don't we add all units from all directories first and then open the initially_opened_module in the resulting env?

alainfrisch replied:
If we ever make the `initially_opened_module` customizable (currently, it is always "Stdlib" or nothing), the logic would seem a bit ad hoc: modules in the same directory than the initially_opened_module wouldn't be allowed to override sub-modules of that one, but modules in any other -I directory would.  Perhaps not a big deal in practice, though.

> In fact, after this PR we almost want to give meaning to the order in which the various -I and -open flags are passed.

Yes, it would seem quite natural then.  People could explicitly pass the same -I directory several times if needed.

alainfrisch commented on utils/load_path.ml:
```diff
@@ -0,0 +1,82 @@
+  let readdir_compat dir =
+    try
+      Sys.readdir (if dir = "" then Filename.current_dir_name else dir)
+    with _ ->
```
You should match on a specific exception constructor.

alainfrisch commented on utils/load_path.mli:
```diff
@@ -0,0 +1,64 @@
+(** Manangement of include directories.
+
+    This module offers a high level interface to locating files in the
+    load path, which is constructed [-I] command line flags and a few
```
constructed "by"/"from"?

diml replied:
updated

alainfrisch commented on utils/load_path.mli:
```diff
@@ -0,0 +1,64 @@
+    doesn't change during the execution of the compiler.
+*)
+
+(** Add a directory to the load path *)
```
AFAICT, the usual style in the compiler code base is to put docstrings after the component.

diml replied:
updated

alainfrisch reviewed (commented):
I did a first review pass and things look globally good.   Some notes:

- This will slightly change the behavior on case-insensitive file-systems.  With the PR, the casing of filenames (except for the first character) needs to match the one of the module name.  Previously, it was ok to rename Foo.cmi into FOO.CMI.  I suppose this is ok.

- Another change is that, in the toplevel, units which were not yet available when the directory was added won't be usable.  Previously, it was possible to start a toplevel with some -I flags while a build system was still running in the background, and still access resulting units once they are build.  Or, more realistically, to have some ocaml script explicitly calling the compiler and then referring to the resulting units.  This could be addressed by a toplevel directive that explicitly refreshes the view on external units.  Not sure it is worth it.

- Yet another change in the toplevel is that with the current "delayed" resolution mechanism, load paths are interpreted relative to the current directory when the lookup is done.  Users could start the toplevel with some -I directives (with relative directories), and explicitly change the current directory at the beginning of the script (e.g. to choose between different implementations/versions).  Again , the change is probably ok.

alainfrisch commented on typing/env.ml:
```diff
@@ -450,14 +450,20 @@ type type_descriptions =
 
+type 'a value_or_persistent =
+  | Value of 'a
+  | Persistent
```
`Persistent` is just a placeholder for remembering that the module is an external unit.  Wouldn't it make sense to keep the actual file name, as discovered during the initial scan?

alainfrisch replied:
Only summaries are serialized, so this would be fine I guess, but it requires a bit more refactoring.

diml replied:
Looking at it, there is `t` in `module_components`, however it is always `empty` when saving a cmi file. I guess we could make it an option, then we would have to bootstrap less often.

alainfrisch commented:
Another small change: if a file is removed during the compilation or during a toplevel session, one could end up, I believe, with read_cmi being passed a non-existing filename, resulting in uncaught exceptions.  Perhaps worth protecting with an explicit error message (or deal gracefully with the situation, as if the file had never been there).

dbuenzli commented:
@diml only a suggestion but since you are in the area would you maybe consider how this could fit with point 2. of [MPR7589](https://caml.inria.fr/mantis/view.php?id=7589)

ghost commented:
Thanks for the review! For the toplevel, yh I agree that a few things will change. Personally, I find the new behavior more intuitive. BTW, you can already rescan a given directory by issuing a new `#directory` primitive. The cost of looking up files in the load path no longer depends on the number of load paths, so it's not really a problem to have duplicates.

> Another small change: if a file is removed during the compilation or during a toplevel session, one could end up, I believe, with read_cmi being passed a non-existing filename, resulting in uncaught exceptions. Perhaps worth protecting with an explicit error message (or deal gracefully with the situation, as if the file had never been there).

Indeed. I guess we can do a `Sys.file_exists` in `Load_path.find`. I suggest to do it only when `Sys.interactive` is `true`, as otherwise it's one more system call that will most of the time not be necessary as the build system should make sure this doesn't happen.

@dbuenzli I believe the information is already available in `Symtable`, so I guess it's just a matter of exposing it in `Toploop`.

alainfrisch commented:
> Indeed. I guess we can do a Sys.file_exists in Load_path.find. 

Or just catch the exception?

ghost commented:
This would work but requires a bit more refactoring. Basically we need to catch it wherever we call `Load_path.find`. I changed all the `Misc.find_in_path` given that they were all of the form `Misc.find_in_path x !Config.load_path`.

ghost commented:
BTW, currently I'm still debugging a build error. The following doesn't type, saying that the value is unbound:

```ocaml
let x = Stdlib.Seq.empty
```

However, the following does type, which is a bit confusing:

```ocaml
module X = Stdlib.Seq
let x = X.empty
```

ghost commented:
We had a look with @lpw25, the problem was that module aliases coming from cmi files where resolved in an empty environment. I've restored the fallback to 2, which solves the issue.

alainfrisch commented:
> We had a look with @lpw25, the problem was that module aliases coming from cmi files where resolved in an empty environment. I've restored the fallback to 2, which solves the issue.

I did not try to understand where this comes from, but my intuitive reaction is that this does not look right.  Can you give some hints on why aliases are resolved in an empty environment, and why the fix should not be to resolve them instead in another environment?

lpw25 commented:
I can't comment on the fix because I haven't read it, but the problem itself is fairly simple. The components value created for a loaded cmi file requires an environment. This environment needs to contain all free identifiers in the cmi file so that they can be looked up. Before this PR an empty environment was sufficient for this since it implicitly contained all persistent identifiers available from the include directories -- and any free identifier in a cmi file must be a persistent identifier[1]. With this PR an empty environment is genuinely empty and so is not sufficient for this purpose.

[1]: Actually, it could also be an identifier from Predef, but I guess we never happen to look one of these up in the environment for the components value. So it would probably be better to use the initial environment rather than an empty environment anyway.

alainfrisch commented:
> Before this PR an empty environment was sufficient for this since it implicitly contained all persistent identifiers available from the include directories 

But then why not simply use the initial environment instead of the empty one?

ghost commented:
Looking at this again, building an environment that only contain the names in `cmi_crcs` should be enough

ghost commented:
I did this and went a bit further. I also had to fix `Matching.get_mod_field`. It feels like this function could be implemented in a more straightforward way, i.e. without constructing a proper environment.

ghost commented:
BTW, for some reason `make bootstrap` tries to build the debugger and a bunch of other stuff. This makes bootstrapping more painful than it should...

alainfrisch commented:
Yes, `make boostrap` does the actual bootstrapping (`make coreboot`) followed by a `make all` to make sure everything's ok.

ghost commented:
Ah, I see. Thanks

jeremiedimino committed:
Refactor load path management and initial environment

- Add a Load_path module which caches files lookup

- Instead of falling back to the external environment, allow to
  declare in the environment that a module comes from the external
  world. This allows persistent structures to shadows non-persistent
  ones

ghost commented:
I implemented `#remove_directory` and fixed a few other issues. The testsuite now passes. I had to add another hook in `Toploop` to implement the self-contained-toplevel test.

This PR is now ready for review.

trefis commented on toplevel/toploop.mli:
```diff
@@ -138,10 +138,25 @@ val print_out_phrase :
 
+type event =
+  | Startup
+  | After_setup
```
That seems somewhat unrelated to the rest of the PR... What did I miss?

Why was it added?

dra27 replied:
It looks like it becomes necessary for self-contained toplevels?

ghost commented:
It was to fix the selt-contained-toplevel test: the names now needs to be added to `Toploop.toplevel_env` as well, and this needs to be done after the setup steps as they reset `Toploop.toplevel_env`. I suppose another way would be to add a dedicated API for that in toploop, i.e.:

```ocaml
val register_module : name:string -> Env.Persistent_signature.t Lazy.t -> unit
```

which would take care of all this.

damiendoligez commented:
>  I guess we can do a `Sys.file_exists` in `Load_path.find`.

Another process might remove the file right after you called `Sys.file_exists`. The only way to be correct is to catch the `Sys_error` when you call `open`.

ghost commented:
Indeed, however `Sys_error` exceptions are hard to match on. All we can do is catch them all, which is not great. For instance, in this case we might want to distinguish `ENOENT` errors from other ones.

dra27 commented on toplevel/toploop.ml:
```diff
@@ -140,10 +139,25 @@ let parse_mod_use_file name lb =
 
+type event =
+  | Startup
+  | After_setup
```
Would this be better as a polymorphic variant, or possibly extensible type, for (slightly) easier adding of future events?

diml replied:
what would we gain by using a polymorphic variant? I'm not a big fan of them in general, except if you use their specific features

dra27 replied:
It's a type which may gain constructors in future versions, so I tend to feel that using one of the type mechanisms which reflects that is more appropriate (so either an actual extensible type or taking advantage of the [> ] part of polymorphic variants).

diml replied:
I suppose yh. Let's use an open type, I don't really like over-using polymorphic variants as they produce awful error messages.

dra27 commented on toplevel/toploop.ml:
```diff
@@ -140,10 +139,25 @@ let parse_mod_use_file name lb =
    ]
 
-(* Hooks for initialization *)
+(* Hook for initialization *)
```
All of this needs to go to opttoploop as well

diml replied:
Indeed, I'll wait until this is approved and then sync the native toplevel in order to cut down the amount of intermediate work.

dra27 reviewed (commented):
Just a few comments - I haven't attempted to review the Env changes.

ghost commented:
Appart from @dra27's comments, is there any other comment left to address?

It seems to me that not much can be done regarding exceptions as catching all `Sys_error` exceptions wouldn't be acceptable. Indeed, it would silently hide real problems such as invalid access permissions.

dra27 commented:
@diml - you could repeat the `Sys.file_exists` in the exception handler in order to know whether to reraise the exception or act as though the first `Sys.file_exists` had failed. (OK, there is a strange scenario where someone deletes the file between the first `file_exists` and the `open` and then recreates it just before the exception handler starts, but I think we can ignore that!)

ghost commented:
I suppose yh, I'll give that a try

damiendoligez commented:
I didn't mean my comment to block this PR. If we can't do better than using `file_exists` then so be it. Simultaneous removal of files by other processes is pretty unusual anyway.

ghost commented:
Ok, I prefer to leave the code as it if nobody minds (except for the toplevel API, I'm happy to follow @dra27 sugestion). What's the deadline to finish this PR if we want to have it in 4.08? It looks like the rebase is going to be quite painful so I need to find a block of time to work on it.

trefis commented:
I think the rebase should be OK actually, the only conflicts you'll have will be with #1610 (which we can look at with Leo, but I don't think it'd be anything major).

ghost commented:
Ack. Actually I exchanged this PR for a Dune PR with @dra27, so David is going to look at finishing this PR.

damiendoligez commented:
>What's the deadline to finish this PR if we want to have it in 4.08?

Something like ASAP... Now that autoconf is merged, there are very few things blocking the freeze and this is one of them.

damiendoligez commented:
@dra27 can we get this PR merged this week?

alainfrisch commented:
Any hope to have this PR in 4.08?  It removes a good part of the big slowdown on compilation time we got since 4.07.

dra27 commented:
Getting there - sorry, bit of a backlog of PRs from me...

dra27 commented:
Rebased - @alainfrisch, would you mind re-running your benchmark test on this revised version? It's also going through [precheck](https://ci.inria.fr/ocaml/job/precheck/183/).

Still to address the review points and also add a test for the original case-sensitivity issue.

alainfrisch commented:
Here are the results:

Branch    |  ocamlc  | ocamlc.opt | ocamlopt.opt |
----------|----------|------------|--------------|
4.06      |  9.79 (0.90) | 3.08 (0.71) | 10.11 (4.73) |
4.07      |  28.55 (18.37) | 21.44 (18.45) | 28.34 (22.31) |
trunk     |  30.50 (19.76) | 22.28 (19.59) | 29.72 (23.67) |
pr2041    |  13.31 (0.93) | 4.30 (1.14) | 11.43 (5.28) |


Each cell gives the real time (sys time in parentheses) in seconds, as returned by `time -p`.

The test consists in compiling a good part of the compiler code base (taken from OCaml 4.04) with a single invocation of the compiler for each tested version (passing all .mli and .ml files at once).  This was run on WSL (Linux subsystem on top of Windows);  flambda disabled; default options passed to the compiler (in particular, no `-principal`).  I did only one run, so while the order of magnitude is interesting, the actual numbers should be taken with a grain of salt. 

Quick conclusion: a good part of the slowdown introduced in 4.07 goes away (but an extra, perhaps unrelated, slowdown from 4.07 to trunk remains).

dra27 committed:
Add test for GPR#1696

dra27 committed:
Add test for MPR#7841

dra27 commented:
Assuming CI passes (I've put it through [precheck](https://ci.inria.fr/ocaml/job/precheck/184/) again), this is ready for final review.

dra27 commented:
And here goes [mark 2](https://ci.inria.fr/ocaml/job/precheck/185/)!

alainfrisch commented:
@dra27 : can you formally approve this PR? Or if you want to have a second pair of eye looking at your contribution, perhaps @diml can have a final look?

ghost commented:
I just had a final look. LGTM

nojb commented:
Will this be cherry-picked to 4.08?

ghost commented:
Yes, that was the plan. I'm doing the cherry-picking now.

ghost commented:
It's done

nojb commented:
Thanks!

ghost commented:
While testing trunk, I realised that we made an incorrect assumption in `acknowledge_pers_struct` in this PR: we assumed that the only global names that are referenced in a cmi file are the ones listed in the imports. However, that's not always the case. For instance:

```
$ echo 'module X = Y' > alias.ml
$ echo -e 'open Alias\nmodule Z = X' > foo.ml
$ ocamlc -no-alias-deps -w -49 -c alias.ml
$ ocamlc -no-alias-deps -w -49 -c foo.ml
$ ocamlobjinfo foo.cmi|grep Y
<no match>
```

Here, `Y` is not counted as an import of `Foo` because it comes from an alias. If we replace `X` by `Y` in foo.ml, the `Y` becomes an import of `Foo`.

I don't know this part of the compiler enough to know whether this is a bug or not. I do find this difference quite surprising though.

In any case, this is an issue as it breaks the build of all projects relying on module aliases to namespace libraries. A simple fix is to just assume that when we get a `Not_found` exception instead of the value `Persistent`, then the looked up module is in fact a persistent module.

nojb commented:
> ```
> env <- add_dir(env, stdlib_dir)
> env <- open(env, Stdlib)
> ...
> ```

I suspect this PR made it so that any module in `stdlib_dir` with the same name as a submodule of `Stdlib` is hidden (eg `Bigarray`). See ocaml/dune#1845. The fix seems to be to add an implicit `-I <stdlib_dir>` on the command line.

ghost commented:
Indeed, the fix should be simple. I'll have a look

ghost commented:
Fixed in #2256

shindere commented:
Currently the 'test_unix_cmdline.byte.output' is broken on Cygwin.

Bisecting seems to show that it is this GPR and in particular commit
7e0862a21290ba9462ef3bd7f82fc6835b25852b which introduced the failure.

Both bytecode and native tests fail when the compiled program is run for the
first time. The native version times out while the bytecode version produces
the following output:

```
Fatal error: no bytecode file specified
/path/to//test_unix_cmdline.byte: line 2: syntax error near unexpected token `$'\001\001C''
/path/to/test_unix_cmdline.byte: line 2: `TßW%.7@IR[gt}c(C'
```

@diml any hint about what could have gone wrong?

Anything I may do to help debugging this?

ghost commented:
@shindere seeing the compilation/execution log could help

shindere commented:
Jérémie Dimino (2019/02/24 23:55 -0800):
> @shindere seeing the compilation/execution log could help

Well no, not really actually :( the logs are totally similar until the
run whose output has already been shown here.

ghost commented:
What program is printing `Fatal error: no bytecode file specified`?

shindere commented:
Jérémie Dimino (2019/02/25 00:05 -0800):
> What program is printing `Fatal error: no bytecode file specified`?

As far as I can tell it's `test_unix_cmdline.byte`.

ghost commented:
I read the code of the test and I can't see anything special.

My only wild guess is that this PR somehow causes the wrong file to be linked in (is it even possible to happen silently?)

shindere commented:
I odn't know really. I'm just re-testing that commit and the one just
before. What's odd is that the produced executable starts by a reference
to ocamlrun, with no #! in front, whereas in the commit before, the
executable is a Windows executable with MZ at the beginning.

@diml do you want me to send you the files produced by the two
invocations of ocamltest, before and after the commit?

ghost commented:
The header for non-custom bytecode files is read from the `camlheader` that is looked up in the load path. One difference between the previous commit and this one is that it is now looked up in a case sensitive way. Maybe that's the reason somehow?

> @diml do you want me to send you the files produced by the two
invocations of ocamltest, before and after the commit?

Sure

shindere commented:
OK so the difference between the behaviours before and after
7e0862a21290ba9462ef3bd7f82fc6835b25852b is in `bytecomp/bytelink.ml`.

Before the commit, the camlheader file is found and copied to the output.
After the commit, the "try" block raises an exception which is then silently
ignored.

ghost commented:
Where is the camlheader file during the execution of tests? Is it possible that it is in a directory `d` such as `Sys.readdir d` fails, but `Sys.file_exists (Filename.concat d "camlheader")` succeeds?

That would explain the difference of behaviours.

dra27 commented:
I'm just looking at this - I have a strong hunch that it's because of Cygwin's `.exe` shim handling. Before, I think it would have attempted to open `camlheader` and Cygwin would spot that `camlheader.exe` exists and open it. Now, we use `Sys.readdir` which will I think be returning `camlheader.exe`

ghost commented:
Oh, are we actually calling this file camlheader.exe on Windows?

dra27 commented:
Specifically Cygwin (the native Windows builds don't add `.exe` for these) - and possibly specific to the build system. One of the joys of Cygwin is not its speed... I'm still waiting for the build to finish :)

ghost commented:
I see. So the fix is to lookup `camlheader.exe` rather than `camlheader` on Cygwin?

shindere commented:
By the way: one problem that still needs to be understood is that the
native version of the same test also fails, but in a different way: the
test times hangs and Jenkins finally marks it as aborted after 20
minutes. As far as I can tell, the problem appears at the same commit
than for the bytecode test.

dra27 commented:
The fix is that the file should never have the `.exe` - there's a parallelism-related bug which I'd noticed there before, so I'm working on a fix for both.

@shindere - I only glanced at the test, but the native version invokes the incorrectly-built bytecode program which seems to be the cause of the hang. When I renamed `stdlib/camlheader_ur.exe` to just `camlheader_ur` those tests all start working again.

The precheck log from this build hasn't been saved (I didn't realise they got purged) - I'm guessing that I ignored the Cygwin failure as being transient, so sorry for not checking that more thoroughly.

shindere commented:
All this sounds great @dra27! If you submit a PR I'll review it promptly
and then I think CI will work again (after the rebase has been disabled
on Cygwin 64), which will be great and help us to catch the real errors.
